### ISSUE TITLE
OCELOT Model and Prediction Evaluation

### ISSUE THREAD

**Issue by mlaurako** [NOAA-EMC/ocelot#92]
This issue tracks the evaluation of OCELOT model predictions against GFS realtime outputs, using surface observations from the zarr file (the same ground truth used in OCELOT model training/inference) as the reference.

---

**Setup:**
- **Ground truth:** Surface observational data from the OCELOT zarr file
- **Model under evaluation:** OCELOT predictions (model: `/scratch3/NCEPDEV/da/Mu-Chieh.Ko/OCELOT/DEV/target-directProj-clean/ocelot/gnn_model/checkpoints/'gnn-epoch-epoch=151-val_loss-val_loss=0.02.ckpt'`)
- **Baseline comparison:** GFS realtime (0.25°), specifically `TMP:2 m above ground` for surface temperature
- **Time period:** March 2025 (init times: 00Z and 12Z; forecast hours: 006 and 012)

---

**Preliminary Results — Init: 2025030100, FHR: 012, Variable: Surface Temperature**

**OCELOT − GFS (mesh grid):**

<img width="1365" height="379" alt="Image" src="https://github.com/user-attachments/assets/fadbfbed-4fa8-4aa9-a2c2-7162a6ed697a" />

**GFS − Surface Obs (obs location):**

<img width="2946" height="770" alt="Image" src="https://github.com/user-attachments/assets/648b9133-e533-4638-8e83-7e7264297270" />

**OCELOT − Surface Obs (obs location):**

<img width="2946" height="772" alt="Image" src="https://github.com/user-attachments/assets/8fd4dd06-c5a2-4441-a80b-ac6acae8c18f" />

The warm bias of OCELOT relative to GFS over the US is consistent with GFS being too cold compared to surface observations. OCELOT is therefore performing better in this regard.

**Comment by mlaurako** [comment]
<hr>
<p><strong>GFS Verification Run Directory:</strong>
<code>/scratch3/NCEPDEV/da/Mu-Chieh.Ko/OCELOT/DEV/target-directProj-clean/ocelot/gnn_model/verif-GFS</code></p>
<hr>
<p><strong>Main Scripts</strong>
Access obs space location from the OCELOT outputs, replace numbers with GFS forecast:</p>

Script | Description
-- | --
make_gfs_true_csv.py | Replace OCELOT predictions with GFS predictions
make_ocelot_gfs_csv.py | Replace ground truth with GFS predictions


<hr>
<p><strong>Source Data</strong></p>
<ul>
<li><strong>GFS:</strong> <code>/scratch3/NCEPDEV/da/Mu-Chieh.Ko/JEDI-nudging/gfs-rt25/${date}</code></li>
<li><strong>OCELOT output:</strong> <code>/scratch3/NCEPDEV/da/Mu-Chieh.Ko/OCELOT/DEV/target-directProj-clean/ocelot/gnn_model/predictions/pred_csv/non-target</code></li>
</ul></body></html>

**Comment by mlaurako** [comment]
### Comparing results between my previous branch (PR#90) and the main branch pulled on Mar 18.

**Main branch settings:**
```
--data_path /scratch4/NAGAPE/gpu-ai4wp/Ronald.McLaren/ocelot/data/v7
--train_start_date 2022-01-01
--train_end_date 2024-11-01
--val_start_date 2024-11-01
--val_end_date 2025-01-01
--train_window_days 12
--val_window_days 8
--val_mode sequential
--val_stride_days 8
--val_update_every_n_epochs 400
--num_layers 16
--lr 5e-4
--weight_decay 1e-5
```
- Early stopping enabled
- training ran until convergence

**PR#90 settings:**
```
data_path: /scratch4/NAGAPE/gpu-ai4wp/Ronald.McLaren/ocelot/data/v6
train_dates: 2022–2024
full_start_date: 2024-01-01
full_end_date: 2024-12-31
train_window_days: 12
val_window_days: 8
train_val_split: 0.9 (90% train, 10% val)
hidden_dim: 128
num_layers: 10
lr: 5e-4
max_epochs: 160
batch_size: 1
weight_decay: 1e-5 (default)
```
- Training data was from 2022 to 2024; annual data rotated every 40 epochs
- random seed
- early stopping enabled
- training ran until convergence

---

The main branch performance is unsatisfactory. Note that results are not directly comparable to PR#90 due to differences in model size and data configuration, though both runs used the same early stopping setup and reached convergence. I will run an additional test on the up-to-date main branch with `hidden_dim=128` and `num_layers=10` to match the PR#90 configuration.

---
Prediction Result (obs-space):
Main:
<img width="2946" height="772" alt="Image" src="https://github.com/user-attachments/assets/2f8ef4e2-fae6-41c5-91f8-5b085d0ecf4e" />
PR#90:

<img width="2946" height="772" alt="Image" src="https://github.com/user-attachments/assets/6c2459ac-1973-4cb2-8217-8addd2541337" />

Prediction Result (mesh):
Main:
<img width="500" height="400" alt="Image" src="https://github.com/user-attachments/assets/06beee4d-30c5-4d48-926e-5ff2d36fb6d8" />
PR #90:
<img width="500" height="400" alt="Image" src="https://github.com/user-attachments/assets/8e913964-05a6-43a5-a449-ea9008728055" />

---

Validation Result (please note that due to the model setup, it's showing different cycles):
Main:
<img width="2946" height="772" alt="Image" src="https://github.com/user-attachments/assets/574d4669-5c86-43fb-a392-2441c0282e39" />

PR #90:
<img width="2946" height="772" alt="Image" src="https://github.com/user-attachments/assets/e6227e37-c96e-4394-a5da-0971961fae75" />

**Comment by mlaurako** [comment]
Model used: the current `main` branch
Checkpoint: `/scratch3/NCEPDEV/da/Azadeh.Gholoubi/gnn-epoch-epoch=1729-val_loss-val_loss=0.10.ckpt`

I ran `run_pred.sh` (updated dataset to `v7`) to generate pred_csv in `/scratch3/NCEPDEV/da/Mu-Chieh.Ko/OCELOT/MAIN/main260330/ocelot/gnn_model/predictions/ep1729/pred_csv/`
Then, use `run_evaluation.sh` to produce the plots in `/scratch3/NCEPDEV/da/Mu-Chieh.Ko/OCELOT/MAIN/main260330/ocelot/gnn_model/evaluation/figures/ep1729`

<img width="750" height="250" alt="Image" src="https://github.com/user-attachments/assets/d760c53f-5126-4e08-98ae-f3d9eb1dd0c4" />

<img width="500" height="400" alt="Image" src="https://github.com/user-attachments/assets/d640eacf-08a4-4fc3-87a6-af608d0779fc" />

<img width="500" height="400" alt="Image" src="https://github.com/user-attachments/assets/53d8d5ab-692e-4ad5-a3d6-6bb12238e438" />

**Comment by mlaurako** [comment]
Observations so far:

1. I cannot duplicate the validation results that Azadeh presented using her checkpoint. Should I further investigate `run_pred.sh` and `predict_gnn.py`? (Azadeh: compare `run_pred.sh` with the version in `evaluation/script`.)
2. In the `EPOCH 1296` performance check, I was able to recreate the figures Azadeh shared from her CSV files. However, even with these results, details are still missing compared to the results from PR90. Further checking is required.
3. The `bipartite` component was not changed between PR90 and PR97 (main 0318), but the degradation was observed since the merge of PR97. Maybe put this aside for now?
4. I tested 'pad' versus 'project' for embedding; the performance was similar. This suggests that the embedding is likely not the reason for the degradation.

**Comment by azadeh-gh** [comment]
@mlaurako here is the path to the new csv files: `/scratch3/NCEPDEV/da/Azadeh.Gholoubi/Rand_MSELoss_Con_surface_obs_epoch913/pred_csv/mesh-grid/`

Could you please generate the analysis comparison plots?

**Below are the GFS  comparison for this checkpoint:**
<img width="2945" height="795" alt="Image" src="https://github.com/user-attachments/assets/5d39ff8c-60cd-45f6-b7a4-69a4b0e7cfbf" />

<img width="2945" height="795" alt="Image" src="https://github.com/user-attachments/assets/650b3d7d-99da-472b-aa76-2c3939a2a54e" />

**Comment by mlaurako** [comment]
Original code:
surface obs only

<img width="2673" height="849" alt="Image" src="https://github.com/user-attachments/assets/ccc7a994-10e8-4963-af98-24ce9d74c157" />

<img width="2945" height="795" alt="Image" src="https://github.com/user-attachments/assets/131b8aa8-ff15-4729-9a7c-457cea679bd2" />

Updated:
surface obs

<img width="2673" height="849" alt="Image" src="https://github.com/user-attachments/assets/12c07936-7c11-481d-bed4-74d5dd3c3ca7" />

<img width="2950" height="1520" alt="Image" src="https://github.com/user-attachments/assets/8be55b8f-0d56-4607-9a5b-5833e759b4b5" />

radiosonde

<img width="2673" height="847" alt="Image" src="https://github.com/user-attachments/assets/f8093f5e-f611-40ef-a672-f6350a7b6ea5" />

<img width="2951" height="1522" alt="Image" src="https://github.com/user-attachments/assets/188593f9-6c96-41a7-a171-2b997786e698" />
